src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java

```diff
@@ -21,85 +22,93 @@
+        LOGGER.fine("Changing workspace selector to '" + StringUtil.singleLine(selector) + "'");
+        WorkspaceManager.setSelector(tool, workspacePath, selector);
+        // Setting workspace selector triggers workspace update
+        // WorkspaceManager.updateWorkspace(tool, workspace.getPath());
 
         return workspace;
     }
 
-    private static void cleanOldWorkspacesIfNeeded(
-            PlasticTool tool,
-            FilePath workspacePath,
-            boolean shouldUseUpdate,
-            List<Workspace> workspaces) throws IOException, InterruptedException {
+    private static boolean mustUpdateSelector(PlasticTool tool, FilePath workspacePath, String selector)
+            throws IOException, InterruptedException, ParseException {
+        String actualSelector = StringUtil.removeNewLines(WorkspaceManager.getSelector(tool, workspacePath));
+        String expectedSelector = StringUtil.removeNewLines(selector);
+        return !actualSelector.equals(expectedSelector);
+    }
 
-        // handle situation where workspace exists in parent path
-        Workspace parentWorkspace = findWorkspaceByPath(workspaces, workspacePath.getParent());
-        if (parentWorkspace != null) {
-            deleteWorkspace(tool, parentWorkspace, workspaces);
-        }
+    private static void deleteOldWorkspacesIfNeeded(PlasticTool tool, FilePath workspacePath, CleanupMethod cleanup, List<Workspace> workspaces)
+            throws IOException, InterruptedException {
 
-        // handle situation where workspace exists in child path
-        List<Workspace> nestedWorkspaces = findWorkspacesInsidePath(workspaces, workspacePath);
-        for (Workspace workspace : nestedWorkspaces) {
+        // Handle situation where workspace exists in child path.
+        List<Workspace> innerWorkspaces = findWorkspacesInsidePath(workspacePath, workspaces);
+        for (Workspace workspace : innerWorkspaces) {
             deleteWorkspace(tool, workspace, workspaces);
         }
 
-        if (shouldUseUpdate) {
-            return;
+        // Handle situation where workspace exists in parent path.
+        List<Workspace> outerWorkspaces = findWorkspacesOutsidePath(workspacePath, workspaces);
+        for (Workspace workspace : outerWorkspaces) {
+            deleteWorkspace(tool, workspace, workspaces);
         }
 
-        Workspace workspace = findWorkspaceByPath(workspaces, workspacePath);
-        if (workspace != null) {
-            deleteWorkspace(tool, workspace, workspaces);
+        if (cleanup == CleanupMethod.DELETE) {
+            Workspace workspace = findWorkspaceByPath(workspacePath, workspaces);
+            if (workspace != null) {
+                deleteWorkspace(tool, workspace, workspaces);
+            }
         }
     }
 
-    private static boolean isSameWorkspacePath(String actual, String expected) {
+    protected static boolean isSameWorkspacePath(String actual, String expected) {
```

Comment:
Question: Does this really need to be protected?

Reply:
Yes. Please see  `CheckoutActionTest.java`. If it is private then I will not be able to access this method and do unit tests. The same goes for other methods that are under tests for this class.

It could be also without modifier (the package scope) however I prefer to always explicitly specify the accessibility scope as it makes it easier to notice that the method/field is accessed somewhere else.